### ISSUE TITLE
Parent project options are being cancelled

### ISSUE THREAD

**Issue by alandefreitas** [cpm-cmake/CPM.cmake#222]
If the parent project has an option that conflicts with the child project, the child project is messing with the scope of the parent project.

For instance

```cmake
message("BUILD_EXAMPLES=${BUILD_EXAMPLES}")
CPMAddPackage(
        NAME pareto
        GIT_REPOSITORY https://github.com/alandefreitas/pareto.git
        GIT_TAG v1.2.0
        OPTIONS
        "BUILD_MATPLOT_TARGETS OFF"
        "BUILD_EXAMPLES OFF"
)
add_executable(spatial_containers spatial_containers.cpp)
target_link_libraries(spatial_containers PUBLIC pareto)
message("BUILD_EXAMPLES=${BUILD_EXAMPLES}")
```

gives me 

```cmake
BUILD_EXAMPLES=ON
-- CPM: adding package pareto@1.2.0 (v1.2.0)
-- Looking for min
-- Looking for min - not found
BUILD_EXAMPLES=OFF
```

which is a problem because some option names are quite common. Some projects create have prefixes in their options but most projects don't.

Of course, people can manually do 

```cmake
set(BUILD_EXAMPLES_PREV ${BUILD_EXAMPLES})
message("BUILD_EXAMPLES=${BUILD_EXAMPLES}")
CPMAddPackage(
        NAME pareto
        GIT_REPOSITORY https://github.com/alandefreitas/pareto.git
        GIT_TAG v1.2.0
        OPTIONS
        "BUILD_EXAMPLES OFF"
)
add_executable(spatial_containers spatial_containers.cpp)
target_link_libraries(spatial_containers PUBLIC pareto)
set(BUILD_EXAMPLES ${BUILD_EXAMPLES_PREV} CACHE BOOL "" FORCE)
message("BUILD_EXAMPLES=${BUILD_EXAMPLES}")
```

for all options but that defeats the purpose of the OPTIONS arguments, which is to do what we cannot already do with FetchContent.

**Comment by TheLartians** [comment]
Yeah, the advantage of options are currently mainly for convenience and error-checking in case a dependency is added twice with incompatible options. It currently doesn't offer protection against naming conflicts with other packages that don't namespace their options (see #180 for similar discussion).

Caching and restoring previous values as in your example may be a good idea to prevent this and shouldn't be too hard to implement.

**Comment by flagarde** [comment]
Hi,

I tried to solve the problem adding :

```cmake
function(declare_option)
  cmake_parse_arguments(OPTION "" "REPOSITORY;OPTION;VALUE" "" ${ARGN})
  if(NOT DEFINED OPTION_REPOSITORY)
    message(FATAL_ERROR "REPOSITORY argument for declare_option is mandatory")
  endif()
  if(NOT DEFINED OPTION_OPTION)
    message(FATAL_ERROR "OPTION argument for declare_option is mandatory")
  endif()
  if(NOT DEFINED OPTION_VALUE)
    message(STATUS "Value for ${OPTION_OPTION} assumed to be ${BoldGreen}ON${Default}")
    set(OPTION_VALUE ON)
  endif()

  if(NOT DEFINED ${OPTION_REPOSITORY}_${OPTION_OPTION})

    list(FIND "${OPTION_REPOSITORY}_OPTIONS_LIST" ${OPTION_OPTION} OPTION_FOUND)

    if(${OPTION_FOUND} STREQUAL "-1")

      list(INSERT "${OPTION_REPOSITORY}_OPTIONS_LIST" 0 "${OPTION_OPTION}" "${OPTION_VALUE}")
      list(INSERT "${OPTION_REPOSITORY}_OPTIONS" 0 "${OPTION_OPTION} ${OPTION_VALUE}")

    else()

      math(EXPR OPTION_FOUND_PLUS_ONE "${OPTION_FOUND}+1")
      list(REMOVE_AT "${OPTION_REPOSITORY}_OPTIONS_LIST" ${OPTION_FOUND_PLUS_ONE})
      list(INSERT "${OPTION_REPOSITORY}_OPTIONS_LIST" ${OPTION_FOUND_PLUS_ONE} "${OPTION_VALUE}")
      math(EXPR OPTION_FOUND "${OPTION_FOUND}/2")
      list(REMOVE_AT "${OPTION_REPOSITORY}_OPTIONS" ${OPTION_FOUND})
      list(INSERT "${OPTION_REPOSITORY}_OPTIONS" ${OPTION_FOUND} "${OPTION_OPTION} ${OPTION_VALUE}")

    endif()

  else()

    list(FIND "${OPTION_REPOSITORY}_OPTIONS_LIST" ${OPTION_OPTION} OPTION_FOUND)
    if(${OPTION_FOUND} STREQUAL "-1")
      list(INSERT "${OPTION_REPOSITORY}_OPTIONS_LIST" 0 "${OPTION_OPTION}" "${${OPTION_REPOSITORY}_${OPTION_OPTION}}")
      list(INSERT "${OPTION_REPOSITORY}_OPTIONS" 0 "${OPTION_OPTION} ${${OPTION_REPOSITORY}_${OPTION_OPTION}}")
    else()
      math(EXPR OPTION_FOUND_PLUS_ONE "${OPTION_FOUND}+1")
      list(REMOVE_AT "${OPTION_REPOSITORY}_OPTIONS_LIST" ${OPTION_FOUND_PLUS_ONE})
      list(INSERT "${OPTION_REPOSITORY}_OPTIONS_LIST" ${OPTION_FOUND_PLUS_ONE} "${${OPTION_REPOSITORY}_${OPTION_OPTION}}")
      math(EXPR OPTION_FOUND "${OPTION_FOUND}/2")
      list(REMOVE_AT "${OPTION_REPOSITORY}_OPTIONS" ${OPTION_FOUND})
      list(INSERT "${OPTION_REPOSITORY}_OPTIONS" ${OPTION_FOUND} "${OPTION_OPTION} ${${OPTION_REPOSITORY}_${OPTION_OPTION}}")
    endif()

  endif()
  set(${OPTION_REPOSITORY}_OPTIONS_LIST ${${OPTION_REPOSITORY}_OPTIONS_LIST} CACHE STRING "Options predefined for ${OPTION_REPOSITORY}" FORCE)
  set(${OPTION_REPOSITORY}_OPTIONS ${${OPTION_REPOSITORY}_OPTIONS} CACHE STRING "Options predefined for ${OPTION_REPOSITORY}" FORCE)
endfunction()

function(print_options)
  cmake_parse_arguments(OPTION "" "REPOSITORY" "" ${ARGN})
  if(NOT DEFINED OPTION_REPOSITORY)
    message(FATAL_ERROR "REPOSITORY argument for declare_option is mandatory")
  endif()
  list(LENGTH ${OPTION_REPOSITORY}_OPTIONS_LIST EMPTY)
  if(${EMPTY} STREQUAL 0)
    message(STATUS "${OPTION_REPOSITORY} doesn't have options")
    return()
  endif()
  string(APPEND RESULT "Options for ${OPTION_REPOSITORY} :${Reset}\n")
  list(LENGTH ${OPTION_REPOSITORY}_OPTIONS_LIST LENGTH)
  math(EXPR LENGTH "${LENGTH}-1")
  set(ITEM "0")
  foreach(OPTION IN LISTS ${OPTION_REPOSITORY}_OPTIONS_LIST)
    math(EXPR MODULO "${ITEM}%2")
    if(${MODULO} STREQUAL "0")
      string(APPEND RESULT "${BoldMagenta}" "   ${OPTION}" " : ")
    else()
      string(APPEND RESULT "${OPTION}" "${Reset}")
      if(NOT ${ITEM} STREQUAL ${LENGTH})
      string(APPEND RESULT "\n")
      endif()
    endif()
    math(EXPR ITEM "${ITEM}+1")
  endforeach()
  message(INFO "${RESULT}")
endfunction()

function(get_option)
  cmake_parse_arguments(OPTION "" "VARIABLE;OPTION;REPOSITORY" "" ${ARGN})
  list(FIND ${OPTION_REPOSITORY}_OPTIONS_LIST ${OPTION_OPTION} FOUND)
  if(${FOUND} STREQUAL "-1")
    set(VALUE "")
  else()
    math(EXPR FOUND "${FOUND}+1")
    list(GET ${OPTION_REPOSITORY}_OPTIONS_LIST ${FOUND} VALUE)
  endif()
  set(${OPTION_VARIABLE} "${VALUE}" PARENT_SCOPE)
endfunction()

function(undeclare_all_options)
  cmake_parse_arguments(OPTION "" "REPOSITORY" "" ${ARGN})
  if(NOT DEFINED OPTION_REPOSITORY)
    message(FATAL_ERROR "REPOSITORY argument for declare_option is mandatory")
  endif()
  set(${OPTION_REPOSITORY}_OPTIONS_LIST "" CACHE STRING "Options predefined for ${OPTION_REPOSITORY}" FORCE)
  set(${OPTION_REPOSITORY}_OPTIONS "" CACHE STRING "Options predefined for ${OPTION_REPOSITORY}" FORCE)
endfunction()

function(undeclare_option)
  cmake_parse_arguments(OPTION "" "REPOSITORY;OPTION" "" ${ARGN})
  if(NOT DEFINED OPTION_REPOSITORY)
    message(FATAL_ERROR "REPOSITORY argument for declare_option is mandatory")
  endif()
  if(NOT DEFINED OPTION_OPTION)
    message(FATAL_ERROR "OPTION argument for declare_option is mandatory")
  endif()
  list(FIND ${OPTION_REPOSITORY}_OPTIONS_LIST ${OPTION_OPTION} OPTION_FOUND)
  if(NOT ${OPTION_FOUND} STREQUAL "-1")
    math(EXPR OPTION_FOUND_PLUS_ONE "${OPTION_FOUND}+1")
    list(REMOVE_AT ${OPTION_REPOSITORY}_OPTIONS_LIST ${OPTION_FOUND} ${OPTION_FOUND_PLUS_ONE})
    math(EXPR OPTION_FOUND "${OPTION_FOUND}/2")
    list(REMOVE_AT ${OPTION_REPOSITORY}_OPTIONS ${OPTION_FOUND})
  endif()
  set(${OPTION_REPOSITORY}_OPTIONS_LIST "${${OPTION_REPOSITORY}_OPTIONS_LIST}" CACHE STRING "Options predefined for ${OPTION_REPOSITORY}" FORCE)
  set(${OPTION_REPOSITORY}_OPTIONS "${${OPTION_REPOSITORY}_OPTIONS}" CACHE STRING "Options predefined for ${OPTION_REPOSITORY}" FORCE)
endfunction()
```
It allows to do 

```cmake
declare_option(REPOSITORY FlakedTuna OPTION CMAKE_CXX_STANDARD VALUE 17)
declare_option(REPOSITORY FlakedTuna OPTION CMAKE_CXX_STANDARD_REQUIRED VALUE TRUE)
declare_option(REPOSITORY FlakedTuna OPTION CMAKE_CXX_EXTENSIONS VALUE FALSE)
declare_option(REPOSITORY FlakedTuna OPTION CMAKE_INSTALL_PREFIX VALUE ${CMAKE_INSTALL_PREFIX})
declare_option(REPOSITORY FlakedTuna OPTION BUILD_TESTS VALUE OFF)
declare_option(REPOSITORY FlakedTuna OPTION BUILD_EXAMPLES VALUE OFF)
print_options(REPOSITORY  FlakedTuna)

CPMFindPackage(NAME FlakedTuna
               GITHUB_REPOSITORY ${FLAKEDTUNA_REPOSITORY}
               GIT_TAG ${FLAKEDTUNA_VERSION}
               FETCHCONTENT_UPDATES_DISCONNECTED ${IS_OFFLINE}
               OPTIONS ${FLAKEDTUNA_OPTIONS})
```

**Comment by alandefreitas** [comment]
@flagarde Where is your code restoring the options from the parent repository?

If the logic happened inside CPMFindPackage, wouldn't it just a matter of (i) for each option, keep the previous value, (ii) do whatever it already does, (iii) restore previous values?

**Comment by flagarde** [comment]
Oh yes I see your point. In fact I did this function to declare the options in a dedicated file. They are all prefixed so easy to unset. Yes you are right we have to keep the previous value first and then turn then back. I could do this in ```declare_option``` and restaure them in ```undeclare_option```.

**Comment by hazelnusse** [comment]
I recently ran across this with a CPMAddPackage() that was setting `OPTIONS "CMAKE_BUILD_TYPE Debug"` and clobbering the parent project settings. I'm not sure why it was doing this in the first place, but I did observe that my settings were getting clobbered. I've reproduced the issue here:

https://github.com/hazelnusse/cmake_cpm

**Comment by flagarde** [comment]
Maybe changing https://github.com/cpm-cmake/CPM.cmake/blob/8e8dcc9a8dbd403118da7a867d3f1c0d43ed922f/cmake/CPM.cmake#L523 to 
```cmake
set(${OPTION_KEY} ${OPTION_VALUE} "" )
``` 
will solve the issue.
Do we really need a CACHEd variable ? If so maybe properties would be better

@hazelnusse  from your code but with a hard copied CPM with the modif : https://github.com/flagarde/cmake_cpm  I obtain : 
```cmake
-- The C compiler identification is GNU 10.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- bar before : 
-- CPM: adding package foo@ (main)
-- foo : Debug;
-- Setting build type to 'RelWithDebInfo' as none was specified.
-- bar after : RelWithDebInfo
-- Configuring done
-- Generating done
```

**Comment by alandefreitas** [comment]
Fetched libraries messing up with OPTIONS, when the intention is to use external libraries as if we had found them with find_package is kind of problematic. You may be "finding the package" locally but it will fail because the user is "fetching the package", or vice-versa. 

So, at worst, it creates an invisible double path that is hard to debug, because you have two different behaviors for the options but you are only really testing one of them, depending on your system, and treating them as somewhat equivalent. On average, it might create bugs that are very hard to find because a subproject might set BUILD_STANDALONE to be OFF (because it's a subproject) while the main project obviously wants BUILD_STANDALONE to be ON, and these names conflict very very very often. It might take a long time for the programmer to identify the mystery why his STANDALONE is not being built. At best, it requires patterns that are ugly and not less verbose than doing it manually, like the ones in this issue.

The ideal would be to iterate through the current options, make a copy, and restore this copy once the package is fetched. The is not zero-cost, but it seems to be worth it, given how often option names repeat in CMake projects, how few options usually there are, and since this kind of security almost always has higher priority than the cost of caching these values.

**Comment by flagarde** [comment]
@alandefreitas If I understand your point correctly you are saying that there is two different behavior depending on the user wish to turn back to te  "finding the package" locally (the find_package) or the Fetching content behavior ?

Using the find_package behavior you skip the option so of course you could end with to different kind of behavior based on this.

But as far as I understand this switch is mainly a programmer preference and not really a user one. It would be a programmer responsability to check is  build is working with both behavior if he wants too. Otherwise he should advice the user not to switch between the two behavior and warn that if so the user is on is own.

**Comment by hazelnusse** [comment]
I think that it is unreasonable to expect/trust the behavior of a system built and installed library/project to be identical to that of one built as part of your cmake build (except in the limited case of header only libraries where you only depend on official releases that also end up packaged by your system). There are simply too many ways they could be different (compiler, flags, C/C++ library version, project version, to name a few) and resulting difference (perhaps negligible, perhaps not) of the parent project. So I agree with @flagarde this being a programmer preference *and* responsibility to verify.

That said, in the case of using CPMAddPackage(), I do think being able to control the sub-project flags and make them potentially different from the parent, or at least know that the sub-projects default settings won't clobber the parents is valuable. @flagarde your change seems to do exactly what I would want. @flagarde do you know if there might be other corner cases where your modification wouldn't work as desired? @TheLartians what do you think?

**Comment by flagarde** [comment]
@hazelnusse I didn't read all the CPM code so I don't know if there is any corner case. An other problem I see but which could be solved I think (programmer responsability) is the following :
libA depends on libB with options X
libC depends on libB with options Y
you use libA and libC, it would be nice to controle the options in your CMakeLists to tell libA and libC to use options Xand Y or a mix of them fitting your needs. This way libB will be included only once and avoid some problem (if you chose the right options of course).

That why I did the small adds upper in this thread (ugly but it works). It would be nice in CPMAddPackage(FOO) to check if variables FOO_OPTIONS exists and if so, bypass the defaults options and use the ones in FOO_OPTIONS. Each lib could redefine them but the main project would take the precedence.

Programmer of libA could use his default options, libC too but programmer needing both could change them without blaming libA or libC to have chosen, made X or Y decision on libB.

Reading the CPM doc I think the CACHED variable has been used to save the options in case you include the lib manytimes with different options to warn the programmer/user. For this we could use PROPERTIES to save the values and only standards Variables (not cached to feed the sub-project).

I don't know if the change breaks this feature but I think that with the change explained before we could solve a lot of problem and give more freedom/responsabilities to the main project programer. And this change would be retrocompatible.

**Comment by alandefreitas** [comment]
@flagarde @hazelnusse 

I think this diverged significantly from what I initially meant. I have no problem with user responsibility or anything like that. I often use find_package and FetchContents in sequence and I know my responsibility. If CPM wants to just be a wrapper for this kind of behavior, I also understand my responsibility and will deal with that or might choose to just use find_package/FetchContents directly.

The problem here refers to what the library intends to do, and whether this follows or helps good practices. Any library should do what it promises to do and anything else is the user's responsibility. If a statistics library doesn't give me plots, that's ok. That's my responsibility. But if this same library gives me `pow(2,2) == 398`, we have a problem of intentionality and it shouldn't be used at all, because function the names have no intentionality in this library, and this is against good practices, and leads me into errors. 

Consider this pattern:

```cmake
set(BUILD_EXAMPLES_PREV ${BUILD_EXAMPLES})
CPMAddPackage(
        NAME pareto
        GIT_REPOSITORY https://github.com/alandefreitas/pareto.git
        GIT_TAG v1.2.0
        OPTIONS
        "BUILD_EXAMPLES OFF"
)
add_executable(spatial_containers spatial_containers.cpp)
target_link_libraries(spatial_containers PUBLIC pareto)
set(BUILD_EXAMPLES ${BUILD_EXAMPLES_PREV} CACHE BOOL "" FORCE)
```

If this is the kind of behavior *intented* by the library, there are problems of intentionality and good practices here. It's a matter of making These are the two bad practices here:

Intentionality: 

Any new user who thinks this would act as the CMake's missing package manager would not apply this pattern and be deeply disappointed by the results. The intentionality is in accordance with how find_package works, in accordance with how FetchContents works, in accordance with how "a Minimal wrapper for find_package and FetchContents" with a `find_or_fetch_package` (with propers warnings about its dangers in the documentation) works. 

It's definitely not in accordance with how the "CMake's missing package manager" works, with how Conan works, with how cargo works. 

[Expressing intention in code](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rp-what) is a good practice. In this case, the idea of "the CMake's missing package manager" is not behaving as a package manager. It's creating a new dependency for a final syntax is just as verbose as find_package/FetchContents, doesn't make clear what my responsibilities are like find_package/FetchContents does, and might mislead people to create an error that will take a while to identify.

Expanding compiling paths: 

There's also this [idea](https://www.youtube.com/watch?v=sBP17HQAQjk) that creating possibilities and optional requirements in your compilation path is deeply problematic. For each new option, there is a new path to be tested, and the number of paths to test grows exponentially. This is very expensive to test, and many many people don't even test more than a single path. If a library leads you towards that, it shouldn't be used at all. A package manager would do everything in its power not to lead you toward that. A simple wrapper for find_package/FetchContents would not. 

In the end, it has nothing to do with responsibility per see. It's just a decision that needs to be made and only @TheLartians can make a final decision. CPM can either 1) be a wrapper for find_package/FetchContents, announce the library as that instead of the CMake's missing package manager, or 2) move towards being a package manager that still has some properly documented limitations as usual.

**Comment by flagarde** [comment]
@alandefreitas Yes, I think we are mixing two stuffs:
-The feature to CPMAddPackage to fallback to find_package and in this case is programer responsability.
-The feature/bug depending on @TheLartians thinking and philosophy/purpose of CPM. I agree with you @alandefreitas  that the actual behavior is misleading but as you said it's fine if this behavior is documented.
We are discussing here mainly on point 2.

My personnal wish woud be to consider this as a bug and fix it but it's @TheLartians decision. This fix would move CPM more as package manager as explained by @alandefreitas.

**Comment by TheLartians** [comment]
Yeah, if I remember correctly the reason for caching the variable was for checking value consistency on later adds of the same library. I had not considered how many CMake libraries don't namespace their options, so I agree that scoping the changes is a necessary to prevent unwanted conflicts. I currently believe that simply removing the cache and leaving version and nested option consistency to the user (or add a more sophisticated system as @flagarde suggested) is the way to go for now. I'm currently quite limited on time, so I'd really appreciate PRs on those features and we can do a more detailed discussion from there. 

I'm not sure I'm fully following the discussion on the philosophy / purpose of CPM.cmake so I'll try to clarify my perspective. The main intention of CPM.cmake was to encourage C++ developers to use define their dependencies in an exactly specified source version, leading to reproducible build behaviour as opposed to the system dependent behaviour of `find_package` function. The amount of times I've seen monorepos with hard-coded dependencies or git submodules being used as dependency management in production shows that there is a real problem in the community. If the script already qualifies for being a "real Package Manager" or just a function wrapper can be debated, but imo the limitations are clearly illustrated by the readme.

The optional  `find_or_fetch_package` behaviour (activated by enabling `CPM_USE_LOCAL_PACKAGES`) and intended for collaborators preferring external package managers, such as docker, Conan or vcpkg. Here I assume that the users know what they are doing and that `find_package` will resolve the correct dependency versions and configurations if available. This may require better documentation.

There is also the function `CPMFindPackage` that has this behaviour enabled by default. Personally, I dislike the function and am thinking about deprecating it or at least making it less prominent in the docs, as it can easily lead to unexpected behaviour depending on the system's state.

**Comment by flagarde** [comment]
thx for the precisions @TheLartians. By the way I agree a lot with the content of your comment. To fix the bug it would be possible to use normal variables as I did in the upper of the thread and maybe change `CPMRegisterPackage("${CPM_ARGS_NAME}" "${CPM_ARGS_VERSION}")` to keep the options.

Then as we are playing with options (we could open an other issue if you agree) , what do you think about the possibility of bypassing the default options looking for a variable called `${FOO_OPTIONS}` in 
```cmake
CPMAddPackage(
        NAME FOO
        GIT_REPOSITORY https://github.com/alandefreitas/pareto.git
        GIT_TAG v1.2.0
        OPTIONS
        "BUILD_EXAMPLES OFF"
)
```
Add the possibility to print the options like `cpm_print_options_package("FOO" )` would be great too and not so difficult to code.

**Comment by TheLartians** [comment]
Hm I'm uncertain to what use-case the options override solves? Currently, you can already override the full package declaration downstream by calling `CPMAddPackage` or `CPMDeclarePackage`. Feel free to take this to another issue so it doesn't get forgotten once this one is resolved.

**Comment by flagarde** [comment]
@TheLartians I was not aware about this function. That seems to solve some needs but if I understand correctly, You need to redefine the GIT_TAG etc right
? If so being able to bypass the OPTIONS would avoid a lot of redefinitions.

**Comment by flagarde** [comment]
@TheLartians An other use-case I can see is :a project X which use other projects using CPM. If the X project want to change options it needs to get CPM and use it. With the bypassing process it would just need to `set(FOO_OPTIONS "...")` and the callees who are using CPM will use it.

**Comment by TheLartians** [comment]
@flagarde imo adding more possibilities to fine-tune and override the package adding behaviour can result in unexpected and hard to debug behaviour. Also this introduces additional edge-cases, such as how we should handle conflicting parameters or multiple calls for overwriting the options.

I believe that in cases where we would want to modify the options downstream, we should also know where the source resides. In that case simply calling `CPMAddPackage` before with the new options is much more predictable and leads to the desired result, at the cost of being a bit more verbose.

**Comment by TheLartians** [comment]
Fix is released in v0.32.0! 🎉
